x-pack/plugin/analytics/src/main/java/org/elasticsearch/xpack/analytics/stringstats/StringStatsAggregator.java

```diff
@@ -0,0 +1,175 @@
+import org.elasticsearch.common.util.IntArray;
+import org.elasticsearch.common.util.LongArray;
+import org.elasticsearch.index.fielddata.SortedBinaryDocValues;
+import org.elasticsearch.search.DocValueFormat;
+import org.elasticsearch.search.aggregations.Aggregator;
+import org.elasticsearch.search.aggregations.InternalAggregation;
+import org.elasticsearch.search.aggregations.LeafBucketCollector;
+import org.elasticsearch.search.aggregations.LeafBucketCollectorBase;
+import org.elasticsearch.search.aggregations.metrics.MetricsAggregator;
+import org.elasticsearch.search.aggregations.pipeline.PipelineAggregator;
+import org.elasticsearch.search.aggregations.support.ValuesSource;
+import org.elasticsearch.search.internal.SearchContext;
+
+import java.io.IOException;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * Metric aggregator that operates over string and returns statistics such as
+ * minimum length, maximum length, average length, the total Shannon entropy and
+ * probability distribution for each character appearing in all terms.
+ */
+public class StringStatsAggregator extends MetricsAggregator {
+
+    final ValuesSource.Bytes valuesSource;
+    final DocValueFormat format;
+
+    /** Option to show the probability distribution for each character appearing in all terms. */
+    private final boolean showDistribution;
+
+    LongArray count;
+    IntArray minLength;
+    IntArray maxLength;
+    /** Accummulates the total length of all fields. Used for calculate average length and char frequencies. */
+    LongArray totalLength;
+    /** Map that stores the number of occurrences for each character. */
+    Map<Character, LongArray> charOccurrences;
+
+    StringStatsAggregator(String name, boolean showDistribution, ValuesSource.Bytes valuesSource, DocValueFormat format,
+                          SearchContext context, Aggregator parent,
+                          List<PipelineAggregator> pipelineAggregators, Map<String, Object> metaData) throws IOException {
+        super(name, context, parent, pipelineAggregators, metaData);
+        this.showDistribution = showDistribution;
+        this.valuesSource = valuesSource;
+        if (valuesSource != null) {
+            final BigArrays bigArrays = context.bigArrays();
+            count = bigArrays.newLongArray(1, true);
+            totalLength = bigArrays.newLongArray(1, true);
+            minLength = bigArrays.newIntArray(1, false);
+            minLength.fill(0, minLength.size(), Integer.MAX_VALUE);
+            maxLength = bigArrays.newIntArray(1, false);
+            maxLength.fill(0, maxLength.size(), Integer.MIN_VALUE);
+            charOccurrences = new HashMap<>();
+        }
+        this.format = format;
+    }
+
+    @Override
+    public ScoreMode scoreMode() {
+        return valuesSource != null && valuesSource.needsScores() ? ScoreMode.COMPLETE : ScoreMode.COMPLETE_NO_SCORES;
```

Comment:
Please add some parenthesis around the predicate here.  Having to remember that `&&` is higher precedence than `?:` is unnecessary cognitive load, so parenthesis will make it more readable, even if they are technically redundant.